client/server/middleware/profiler.js

```diff
@@ -0,0 +1,72 @@
+const fs = require( 'fs' );
+const path = require( 'path' );
+const v8Profiler = require( 'v8-profiler-next' );
+
+/**
+ * This should be dynamically imported:
+ *
+ * if ( shouldProfile ) {
+ *   app.use ( require( 'calypso/server/middleware/profiler' )() )
+ * }
+ *
+ * to avoid importing v8-profiler-next in production environments.
+ */
+module.exports = () => {
+	let IS_PROFILING = false;
+
+	// Generates a CPU profile compatible with VS Code's viewer.
+	v8Profiler.setGenerateType( 1 );
+
+	const profilesRoot = path.resolve(
+		path.join( __dirname, '../../../profiles', new Date().toISOString() )
+	);
+	try {
+		fs.accessSync( profilesRoot );
+	} catch {
+		fs.mkdirSync( profilesRoot, { recursive: true } );
+	}
+
+	return ( req, res, next ) => {
+		// Avoid profiling certain requests (like for static files) and don't
+		// start profiling if we already are.
+		if (
+			IS_PROFILING ||
```

Comment:
What's the purpose behind `IS_PROFILING` in the first place? To prevent profiling multiple concurrent requests?

Reply:
Right, so if you visit "/themes", it will start profiling. If you then also visit "/home/foo" before the first request finishes, it would start profiling again with a different ID. At the end, you would get two different profiles with overlapping data, since obviously the CPU is shared between the two requests.

Since my main goal was to profile a single request, I figured it would be better to just start the profiler once. (And if you were to visit a second URL concurrently, that data is still included in that first profile, up to the point the first request finishes.)

Reply:
Makes sense. I think it's worth documenting this trait as well to avoid confusion 😉